colossalai/tensor/_ops/linear.py

```diff
@@ -30,32 +30,33 @@ def colo_linear(types, args, kwargs, pg):
 
     # Add communication logic before and after linear call.
     if isinstance(weight, ColoTensor):
-        if weight.shard_spec == None:
+        if weight.shard_spec == None or weight.shard_spec.num_policy == 0:
             if isinstance(input_tensor, ColoTensor):
                 input_tensor = input_tensor.torch_tensor()
             if isinstance(weight, ColoTensor):
                 weight = weight.torch_tensor()
             return torch.nn.functional.linear(input_tensor, weight, bias)
-        elif weight.shard_spec == '1Drow':
-            # Input:S[1] x Weight:S[0] = Output:P
-            # All-Reduce(Output) + bias = res
-            assert divide(input_tensor.shape[-1], gpc.tensor_parallel_size) == weight.size(-1), \
-            'Invalid shapes in 1Drow forward: input={}, weight={}. Expected last dim of input {}.'.format(
-                input_tensor.shape, weight.size, weight.size[-1] * gpc.tensor_parallel_size)
-            # Input:S[1]
-            input_per_partition = split_forward_gather_backward(input_tensor, ParallelMode.PARALLEL_1D, dim=-1)
-            # Output:P
-            device = get_current_device()    # TODO where to put to(deivce)?
-            weight_ = weight.torch_tensor().to(device)
-            partial_output = torch.nn.functional.linear(input_per_partition, weight_)
-            # Reduce(Output)
-            output = reduce_input(partial_output, ParallelMode.PARALLEL_1D)
-            # Bias
-            if bias is not None:
-                bias_ = bias.to(device)
-                output = output + bias_
-            return output
-
+        elif weight.shard_spec.num_policy == 1:
+            if ComputePattern.TP1DRow in weight.shard_spec.compute_patterns:
+                # Input:S[1] x Weight:S[0] = Output:P
+                # All-Reduce(Output) + bias = res
+                assert divide(input_tensor.shape[-1], gpc.tensor_parallel_size) == weight.size(-1), \
+                'Invalid shapes in 1Drow forward: input={}, weight={}. Expected last dim of input {}.'.format(
+                    input_tensor.shape, weight.size, weight.size[-1] * gpc.tensor_parallel_size)
+                # Input:S[1]
+                input_per_partition = split_forward_gather_backward(input_tensor, ParallelMode.PARALLEL_1D, dim=-1)
+                # Output:P
+                weight_ = weight.torch_tensor()
+                partial_output = torch.nn.functional.linear(input_per_partition, weight_)
+                # Reduce(Output)
+                output = reduce_input(partial_output, ParallelMode.PARALLEL_1D)
+                # Bias
+                if bias is not None:
+                    bias_ = bias
+                    output = output + bias_
+                return output
```

Comment:
return a ColoTensor?
do you think it should be a default procedure?

Reply:
I return a ColoTensor now. But I'm not sure whether we shall do that. We have hijacked all Parameters in context(i.e. All torch functions have redirected into ours.), so the type of Input should take no effects in my opition.

Reply:
OK